Rx/v2/src/rxcpp/operators/rx-zip.hpp

```diff
@@ -7,63 +7,130 @@
 
 #include "../rx-includes.hpp"
 
+/*! \file rx-zip.hpp
+
+    \brief Bring by one item from all given observables and select a value to emit from the new observable that is returned.
+
+    \tparam AN  types of scheduler (optional), aggregate function (optional), and source observables
+
+    \param  an  scheduler (optional), aggregation function (optional), and source observables
+
+    \return  Observable that emits the result of combining the items emitted and brought by one from each of the source observables.
+
+    If scheduler is omitted, identity_current_thread is used.
+
+    If aggregation function is omitted, the resulting observable returns tuples of emitted items.
+
+    \sample
+
+    Neither scheduler nor aggregation function are present:
+    \snippet zip.cpp zip sample
+    \snippet output.txt zip sample
+
+    Only scheduler is present:
+    \snippet zip.cpp Coordination zip sample
+    \snippet output.txt Coordination zip sample
+
+    Only aggregation function is present:
+    \snippet zip.cpp Selector zip sample
+    \snippet output.txt Selector zip sample
+
+    Both scheduler and aggregation function are present:
+    \snippet zip.cpp Coordination+Selector zip sample
+    \snippet output.txt Coordination+Selector zip sample
+*/
+
 namespace rxcpp {
 
 namespace operators {
 
 namespace detail {
 
-template<class T>
+template<class Observable>
 struct zip_source_state
 {
+    using value_type = rxu::value_type_t<Observable>;
     zip_source_state() 
         : completed(false) 
     {
     }
-    std::list<T> values;
+    std::list<value_type> values;
     bool completed;
 };
 
 struct values_not_empty {
-    template<class T>
-    bool operator()(zip_source_state<T>& source) const {
+    template<class Observable>
+    bool operator()(zip_source_state<Observable>& source) const {
         return !source.values.empty();
     }
 };
 
 struct source_completed_values_empty {
-    template<class T>
-    bool operator()(zip_source_state<T>& source) const {
+    template<class Observable>
+    bool operator()(zip_source_state<Observable>& source) const {
         return source.completed && source.values.empty();
     }
 };
 
 struct extract_value_front {
-    template<class T>
-    T operator()(zip_source_state<T>& source) const {
+    template<class Observable, class Value = rxu::value_type_t<Observable>>
+    Value operator()(zip_source_state<Observable>& source) const {
         auto val = std::move(source.values.front());
         source.values.pop_front();
         return val;
     }
 };
 
-template<class Coordination, class Selector, class... ObservableN>
-struct zip_traits {
-    typedef std::tuple<ObservableN...> tuple_source_type;
-    typedef std::tuple<zip_source_state<typename ObservableN::value_type>...> tuple_source_values_type;
+template<class... AN>
+struct zip_invalid_arguments {};
```

Comment:
I decided not to since some compilers did not trigger the static assert and this makes the operator and arguments clear in the debug output when the overload is missing